experiments/process_sandbox/filetree.h

```diff
@@ -0,0 +1,319 @@
+       * directory, or a pointer to another instance of this class.
+       */
+      using DirLookupResult =
+        std::variant<std::monostate, platform::handle_t, DirPtr>;
+
+      /**
+       * Look up a directory (either a real directory handle or a pointer to an
+       * instance of this class).  The variant contains `std::monostate` if
+       * the path is not a directory.
+       */
+      DirLookupResult get_dir(const std::string& dir)
+      {
+        auto it = directory.find(dir);
+        if (it == directory.end())
+        {
+          return {};
+        }
+        return std::visit(
+          [](auto&& arg) -> DirLookupResult {
+            using T = std::decay_t<decltype(arg)>;
+            if constexpr (std::is_same_v<T, FileHandle>)
+            {
+              return {};
+            }
+            else if constexpr (std::is_same_v<T, DirPtr>)
+            {
+              return arg;
+            }
+            else if constexpr (std::is_same_v<T, DirHandle>)
+            {
+              return arg.fd;
+            }
+          },
+          it->second);
+      }
+
+      /**
+       * Look up a file in this directory.  Returns the handle to the file if
+       * it exists.
+       */
+      std::optional<platform::handle_t> get_file(const std::string& file)
+      {
+        auto it = directory.find(file);
+        if (it == directory.end())
+        {
+          return {};
+        }
+        if (auto* h = std::get_if<FileHandle>(&it->second))
+        {
+          return h->fd;
+        }
+        return {};
+      }
+
+      /**
+       * Add a child directory tree to this directory.
+       */
+      void add_tree(const std::string& name, DirPtr& tree)
+      {
+        directory[name] = tree;
+      }
+
+      /**
+       * Add a directory tree represented by a directory handle.  Takes
+       * ownership of the handle.
+       */
+      void add_directory(const std::string& name, platform::Handle&& h)
+      {
+        directory[name] = DirHandle(std::move(h));
+      }
+
+      /**
+       * Add a file, represented by a directory handle.  Takes ownership of the
+       * handle.
+       */
+      void add_file(const std::string& name, platform::Handle&& h)
+      {
+        directory[name] = FileHandle(std::move(h));
+      }
+    };
+
+    /**
+     * Pointer to a single directory in the tree.
+     */
+    using DirPtr = ExportedDirectory::DirPtr;
+
+    /**
+     * The root of this exported file tree.
+     */
+    DirPtr root = std::make_shared<ExportedDirectory>();
+
+    /**
+     * Add a handle at a specific page.  This is a helper method for the public
+     * interfaces that add a file or directory handle.  The `add` parameter
+     * contains a callable object that takes the base file name and the
+     * directory pointer as arguments and adds it.
+     *
+     * Returns true on success, false on failure.  This can fail if one of the
+     * path components is a directory handle.
```

Comment:
So `DirHandle` can only be leaf directories? And if you add a sub-dir there, you need to convert to a `DirPtr`?